insecure/corruptible/factory_test.go

```diff
@@ -0,0 +1,225 @@
+package corruptible
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	testifymock "github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+
+	"github.com/onflow/flow-go/insecure"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/libp2p/message"
+	"github.com/onflow/flow-go/network"
+	"github.com/onflow/flow-go/network/codec/cbor"
+	"github.com/onflow/flow-go/network/mocknetwork"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+func TestRegisterAdapter(t *testing.T) {
+	f := NewCorruptibleConduitFactory(flow.BftTestnet, unittest.IdentifierFixture(), cbor.NewCodec())
+
+	adapter := &mocknetwork.Adapter{}
+
+	// registering adapter must go successful
+	require.NoError(t, f.RegisterAdapter(adapter))
+
+	// second attempt on registering adapter must fail
+	require.Error(t, f.RegisterAdapter(adapter))
+}
+
+// TestNewConduit_HappyPath checks when factory has an adapter registered, it can successfully
+// create conduits.
+func TestNewConduit_HappyPath(t *testing.T) {
+	f := NewCorruptibleConduitFactory(flow.BftTestnet, unittest.IdentifierFixture(), cbor.NewCodec())
+	channel := network.Channel("test-channel")
+
+	adapter := &mocknetwork.Adapter{}
+
+	require.NoError(t, f.RegisterAdapter(adapter))
+
+	c, err := f.NewConduit(context.Background(), channel)
+	require.NoError(t, err)
+	require.NotNil(t, c)
+}
+
+// TestNewConduit_MissingAdapter checks when factory does not have an adapter registered,
+// any attempts on creating a conduit fails with an error.
+func TestNewConduit_MissingAdapter(t *testing.T) {
+	f := NewCorruptibleConduitFactory(flow.BftTestnet, unittest.IdentifierFixture(), cbor.NewCodec())
+	channel := network.Channel("test-channel")
+
+	c, err := f.NewConduit(context.Background(), channel)
```

Comment:
Why can't the adapter registration be done when creating a new conduit? That would reduce the chance of forgetting to do this.

Reply:
For the default (i.e., correct) conduits, the `adapter` is registered at [the creation time](https://github.com/onflow/flow-go/blob/9cb2a74ae2cc062efb7faeb817fa3d1831c7c80b/network/p2p/conduit/conduit.go#L48).

The corrupted conduits, however, do not interact directly with `adapter`. Rather all interactions are done through their `master` (i.e., `CorruptibleConduitFactory`). However, I think your question is "Why can't the adapter registration be done when creating a _new conduit factory_ (and not a conduit itself)?" The answer is: 

The instantiation of the [`adapter` (i.e., `network`)](https://github.com/onflow/flow-go/blob/master/network/p2p/network.go#L134) happens after the creation of the `ConduitFactory`. Indeed, the `ConduitFactory` is passed as an instantiation parameter to the `network` factory. 

That is the reason we may not pass `adapter` when creating a new `ConduitFactory`, since it basically does not exist at that time.